### ISSUE TITLE
app-request: Store/Trading

### ISSUE THREAD

**Issue by Gerold55** [Gerold55/minetest-laptop#40]
Eventually we need a place like maybe a web browser with individual places for people to setup their own store (servers only). Also a default trading section for a player to trade gold or mese for an item like diamonds.

Idea collection:
- [ ] Credit card item, readable in computers.
- [ ] Banking app to set up the card (Player assignment + PIN stored on item. The bank account value stored in cloud)
- [ ] money transfer to other player in banking app
- [ ] Appstore to buy apps using (pricing for calculator, mail or sticky notepad?)
- [ ] ATM machine to change money-items from other mods to virtual bank account, or back to "hard value" - Compatible items hard-coded, starting with money:coins items

**Comment by bell07** [comment]
This I do not see in "Normal computers" because you get gold or Mese from laptop.

But maybe additional hardware "cash machine" that can accept money-items to store a "bank balance" in the "cloud" for the player.
Then digital payment: App-Store. At the time all apps are automatically installed on all computers with default launcher. The default launcher can be changed to show only "payed/installed" apps. That means you start with an empty laptop, just the settings and appstore app enabled by default. Then you can buy apps and themes using "Cloud-Money" ;-)

**Comment by Gerold55** [comment]
Yeah that's sorta what I was saying to be able to sell ingots for cash and then deposit it into your bank account. Using the bank account like you had said people can buy different apps items etc etc. Maybe we can have an optional dependency on a currency mod. For the laptop it would be digital currency.

**Comment by bell07** [comment]
googol requested compatibility to minecoins in forum
https://forum.minetest.net/viewtopic.php?f=9&t=18920#p303640
Assume he meant the [BitChange (money)](https://forum.minetest.net/viewtopic.php?f=11&t=7821) mod

**Comment by Gerold55** [comment]
There could be a few optional dependencies for money mods. Like bitcoin or any other one that's used in VanessaE's servers

**Comment by Gerold55** [comment]
We could also make our own way of currency and when people give suggestions add it in as an optional dependency

**Comment by bell07** [comment]
The major question for me is what should be possible on laptop using money. I do not like the idea the laptop or computers interact directly with "hard money" like coins because it is not realistic. I like more the idea to get an new computer "ATM" that is registered hardware and can transfer inserted coins / any other "hard money" items to soft "Bank account value" stored in our "cloud" or able to pay out.

Then the laptop can work with bank account using Banking app. Next ideas:
- [ ] money transfer to other player
- [ ] Appstore to buy apps (pricing for calculator, mail or sticky notepad?)
- [ ] ???

**Comment by Gerold55** [comment]
We could make use parts of the ATM mod and/or integrate it in and use it with the laptop mod or to make it depend on it and have part of the code look for it and you should be able to use your pin to log into your bank account on the laptop via the banking app. Thing is we would have to make it so credit cards would somehow be used and we could make a virtual aspect of minecoins via the ATM and/or app with a MineCoins ATM machine.

**Comment by bell07** [comment]
Ah, ok, credit card item, readable on laptop, if the pin is entered.
I edit you first post and collected all previous suggestions to 1 place. Feel free to enhance the which list

**Comment by bell07** [comment]
https://github.com/Gerold55/minetest-laptop/issues/71#issuecomment-361630452

> Could you come up with a concept for it? I think that we should figure out the currency system also along with it. Without currency the app store will possibly be useless I think.

My concept ends at "A per-player modmeta attribute to store virtual bank account value." and "A banking app to manage the account". Some additional ideas was collected above.

They are some other coins or money mods we need to be compatible. The compatibility means we can exchange the coins/money from other mods to our virtual "laptop inc. bank account" value and back.
I do not use them, therefore no idea from my site.

And additional gameplay related questions are open for me. 

Question 1: Without the additional coins and money mods: How can value be added to the account?

Question 2: Usually money is not appear and disapper but transferred from Account 1 to Account 2. If an app is payed in the App-Store, the value is transferred from Player's Account to ??? Account?

Sorry for more questions then answers from my site

**Comment by Gerold55** [comment]
For currency if no additional mods have been enabled we should use some kind of ore.

**Comment by Gerold55** [comment]
We should try to make as many currency mods as possible to the optional dependencies list and by default use resources to craft or trade in for currency.

**Comment by bell07** [comment]
I think for the first it should be enough to give value to mese, gold and diamond items. The next question how we can convert diamond (=physical object) to account value (virtual). I do not know laptops with integrated "diamond reader"'s oO. I think we need an ATM that is able to pay in and out diamonds gold and mese items

**Comment by Gerold55** [comment]
Yeah how hard would that be to implement?

**Comment by bell07** [comment]
Less then Printer. Hardware + launcher "ATM OS"

**Comment by Gerold55** [comment]
Thanks again for all your help on this 😄

**Comment by Gerold55** [comment]
Oh ok 👍

**Comment by Gerold55** [comment]
![screenshot 42](https://user-images.githubusercontent.com/13843095/35584796-12d08a16-05c4-11e8-9ab4-65e9a95240fd.png)

**Comment by bell07** [comment]
I think we need to define  the Minetest money currency name at the first. BitTestCoins? MTC (Mine test coins)? The value base is connected to gold, so gold_lump is 1

Second we need an exchange table like
```
mtc_rate = {
["default:gold_lump"] = 10,
["default:gold_ingot"] = 10,
["default:mese_crystal"] = 50,
["default:diamond"] = 100,
["othermod:othermoney"] = 12,
["homedecor:coin"] = 0.6,
}
```

This is the most difficult part, to get the table balanced, because it does allow to exchange gold to our currency and then buy diamonds. So if you does a big gold vein, you can buy anything.

At the end the app pricing, as attribute in app definition, like 
```
laptop.register_app("mail", {
	app_name = "Mail",
	app_icon = "laptop_email_letter.png",
	app_info = "Write Mail to Other Players",
	app_groups = { "internet", "office" },
	app_price = 99,95,
....

**Comment by Gerold55** [comment]
Sounds great to me 👍

**Comment by bell07** [comment]
My feel is this all about currency should be outside the laptop mod. It should be an own balanced currency/economy mod with optional dependency to the laptop mod.
At this point we need to check which existing economy/currency mod could be enhanced for laptop's online banking and online payments. I do not like to reinvent the wheel.

**Comment by Gerold55** [comment]
Yeah I'm making a list of currency mods which are either WIP or have been released.

**Comment by Gerold55** [comment]
BitChange
[Mod] Bank Accounts [1.1][bank_accounts]
[Mod] Mint & Automatic Commerce Machines [mint]
[Mod] Item sell/buy system / banking (for servers) [economy]
[Mod] atm -- Bank accounts for the currency mod
[Mod] Automats and currency [bank]

**Comment by bell07** [comment]
CAn you provide URL's to the found mods?

**Comment by Gerold55** [comment]
https://forum.minetest.net/viewtopic.php?f=11&t=7821&hilit=currency

**Comment by Gerold55** [comment]
https://forum.minetest.net/viewtopic.php?f=11&t=7821&hilit=currency

**Comment by Gerold55** [comment]
https://forum.minetest.net/viewtopic.php?f=9&t=16271&hilit=currency

https://forum.minetest.net/viewtopic.php?f=11&t=4111&hilit=currency

**Comment by Gerold55** [comment]
https://forum.minetest.net/viewtopic.php?f=11&t=14101&hilit=currency

https://forum.minetest.net/viewtopic.php?f=9&t=15029&hilit=currency

https://forum.minetest.net/viewtopic.php?f=9&t=13871&hilit=currency

**Comment by Gerold55** [comment]
That's all of the URL's.

**Comment by bell07** [comment]
The "[Mod] Automats and currency [bank]" seems to be an enhancement for https://github.com/Bremaweb/money2

The "money2" mod match our requirements: Per-Player virtual bank account. Without admin rights and laptop, payments are possible using `/money pay playername value`. Technically there is a proper API. Trough "[Mod] Automats and currency [bank]" some additional functionality exist. 

But the decision is not taken, I need to get a deeper look to the other mods too. The "money2"  needs an update away from files-based storage to modmeta or playermeta..

**Comment by bell07** [comment]
My analysis:

bitchange - Items only, no virtual bank account

currency - Items only, no virtual bank account

bank_accounts - based on currency
Seems to be right direction but I did not understand how to use

atm - addition to currency
money is trough "income", means 5 per 100 seconds for all players
3x money2money changer ATM and one transfer ATM. No way to change Items to money

money2 - as described above, proper API for virtual bank accounts
bank - money items and credit card for money2. Unsure how to use
automats - some automats to pay in/out money or items

mint - Dead download link

economy - crash with intllib. Abadoned by developer. He recommends Minegeld (currency)
But basically this meets our requirements.

**Comment by bell07** [comment]
It seems to be 2 different money ecosystems "currency", "money2" in the list. Trough review I like more money2, but the currency ecosystem seems to be more active.  In the currency ecosystem we need to choice between "bank_accounts" and "atm".

So interested on other opinions

**Comment by Gerold55** [comment]
Is it possible to mix them? Or no? Or just use both

**Comment by bell07** [comment]
It is like in real live to support different banks from different countries with different currencies. Of course we can develop an "online-banking" app for each "bank". But the main reason to support the banks is to pay apps. We have two ways: Pricing for each app with 3x currencies or we define exchange rate. and do pricing for one of the both currencies. "bank_accounts" and "atm" does have already the same currency (currency's "Mine-Geld"), the money2 is "cr".

And we need a abstraction layer to work with all banks in the same way ....

**Comment by Gerold55** [comment]
Which one would be easier for you to do? Should we use MineGeld currency because it's used on a lot of servers? If we do our own currency rate how should it be done?

**Comment by bell07** [comment]
I think it is the same effort. Maybe you ask the main developers of the 3x mods (bank_accounts, atm and money2) which one like to support us in integration. Because the mod developers does have the vision about the currency capabilities and can sey how it should be integrated in best way.
Depending on support the mod priorities could be set.

**Comment by Gerold55** [comment]
![screenshot 48](https://user-images.githubusercontent.com/13843095/35684042-b404e06a-0733-11e8-88c1-0268e2cede9f.png)

**Comment by Gerold55** [comment]
![screenshot_2018-02-01-20-58-04](https://user-images.githubusercontent.com/13843095/35763463-c24e2bb6-087a-11e8-9844-27eb8edee364.png)

**Comment by Gerold55** [comment]
@bell07 couldn't we use ores/gems as the default currency temporarily? Then we can add other currencies later on?

**Comment by bell07** [comment]
The issue is not about the currency but which table is used for bank account to store virtual money values. Each ATM mod does have own table.
bank_accounts uses accounts.balance[playername] table for example, the money2 uses per-player files money_" .. name .. ".txt" in the world dir, accesses each time.

Of course we can start own account implementation in the laptop mod, store the values to our "cloud". It is like to start an own bank in real life. This way we need all additions a bank usually have, like ATM's, and other "bank services".
My hope was we can just use already existing bank accounts and just implement the apps payment, and maybe money transfers to other players (Online banking app).

**Comment by Gerold55** [comment]
Maybe we can use bank_accounts or is money2 easier to use?

**Comment by bell07** [comment]
The issue is I do use all of them. For me an appstore is ok if all apps are for free (like usual linux distri repo). So may an other developer should decide

**Comment by Gerold55** [comment]
Which mod promises more? We need to have a bank so if other devs decide to create an app and sell it in game (server) then the option will be there.

**Comment by Gerold55** [comment]
If you want to set it up that way for servers yes and for singleplayer it'll go to an imaginary account in the cloud.

**Comment by Gerold55** [comment]
Where are we on this?

**Comment by bell07** [comment]
Trading app available as separate mod in https://github.com/orbitaldecay/central_bank modpack.

We need a Bank logo that could be used for App-Icon

Looking back to https://github.com/Gerold55/minetest-laptop/issues/40#issuecomment-361927638 the decision is taken.The central_bank new, but trough active development and proper API have the potential be a brige to all money ecosystems in future.
